clang/lib/Sema/SemaDeclCXX.cpp

```diff
@@ -3598,10 +3598,22 @@ Sema::ActOnCXXMemberDeclarator(Scope *S, AccessSpecifier AS, Declarator &D,
   if (getLangOpts().SYCLIsDevice) {
     if (auto Value = dyn_cast<ValueDecl>(Member)) {
       if (isTypeDecoratedWithDeclAttribute<SYCLDeviceGlobalAttr>(
-              Value->getType()) &&
-          Value->getAccess() != AS_public) {
-        Diag(Loc, diag::err_sycl_device_global_not_publicly_accessible)
-            << Value;
+              Value->getType())) {
+        if (Value->getAccess() == AS_private) {
+          Diag(Loc, diag::err_sycl_device_global_not_publicly_accessible)
+              << Value;
+        }
+        const DeclContext *DC = Member->getDeclContext();
+        while (!DC->isTranslationUnit()) {
+          if (auto Decl = dyn_cast<NamedDecl>(DC)) {
+            if (Decl->getAccess() == AS_private) {
```

Comment:
Should this be changed to `Decl->getAccess()!=AS_public `as well ?

Reply:
@elizabethandrews Changing it to !=AS_public here trips up cases where device_global is declared inside a struct (such as L16 in the test), which is clearly not private but there is no explicit access specifier, but the Decl is AS_none for some reason.

Reply:
I changed it to `Decl->getAccess() == AS_private || Decl->getAccess() == AS_protected` so that it can correctly diagnose cases where a device_global is declared as a protected member of a class/struct